### PR TITLE
Add 'label match' support for override and tag rules

### PR DESCRIPTION
**Pull request by chickenlj**
Related to #11091 and #11090

### CODE REVIEW

**Comment by CrazyHZM** [comment]
@chickenlj Please commit this change to the 3.2 branch.

**Comment by chickenlj** [comment]
> @chickenlj Please commit this change to the 3.2 branch.

I have changed the target branch to 3.2

**Review by CrazyHZM — CHANGES_REQUESTED** [review]

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/router/state/AbstractStateRouter.java [hunk 1]
```diff
@@ -83,7 +87,7 @@ public StateRouter<T> getNextRouter() {
 
     @Override
     public void notify(BitList<Invoker<T>> invokers) {
-        // default empty implement
+        this.invokers = invokers;
```
Comment (AlbumenJ) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/router/state/AbstractStateRouter.java:90]: Should not depend on notify here. Some routers will override notify method.

File: dubbo-cluster/src/main/resources/META-INF/dubbo/internal/org.apache.dubbo.rpc.cluster.router.state.StateRouterFactory [hunk 2]
```diff
@@ -2,5 +2,7 @@ tag=org.apache.dubbo.rpc.cluster.router.tag.TagStateRouterFactory
 condition=org.apache.dubbo.rpc.cluster.router.condition.ConditionStateRouterFactory
 service=org.apache.dubbo.rpc.cluster.router.condition.config.ServiceStateRouterFactory
 app=org.apache.dubbo.rpc.cluster.router.condition.config.AppStateRouterFactory
+provider-app=org.apache.dubbo.rpc.cluster.router.condition.config.ProviderAppStateRouterFactory
+script-app=org.apache.dubbo.rpc.cluster.router.script.config.AppScriptRouterFactory
```
Comment (AlbumenJ) [inline dubbo-cluster/src/main/resources/META-INF/dubbo/internal/org.apache.dubbo.rpc.cluster.router.state.StateRouterFactory:6]: Script router should not enabled by default.

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

**Comment by chickenlj** [comment]
```
[08:42:11:011]  INFO utils.SerializeSecurityConfigurator:  [DUBBO] Read serialize blocked list from jar:file:/usr/local/dubbo/app/dependency/dubbo-3.2.0-beta.6-SNAPSHOT.jar!/security/serialize.blockedlist, dubbo version: 3.2.0-beta.6-SNAPSHOT, current host: 172.27.0.3
Tests run: 11, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 4.066 s <<< FAILURE! - in org.apache.dubbo.sample.tri.direct.TriDirectPojoClientTest
testBoxed(org.apache.dubbo.sample.tri.direct.TriDirectPojoClientTest)  Time elapsed: 0.134 s
greetStream(org.apache.dubbo.sample.tri.direct.TriDirectPojoClientTest)  Time elapsed: 0.028 s
greetUnaryResponseVoid(org.apache.dubbo.sample.tri.direct.TriDirectPojoClientTest)  Time elapsed: 0.01 s
unaryFuture(org.apache.dubbo.sample.tri.direct.TriDirectPojoClientTest)  Time elapsed: 0.012 s
greetUnary(org.apache.dubbo.sample.tri.direct.TriDirectPojoClientTest)  Time elapsed: 0.018 s
greetChildPojo(org.apache.dubbo.sample.tri.direct.TriDirectPojoClientTest)  Time elapsed: 0.016 s
greetServerStream(org.apache.dubbo.sample.tri.direct.TriDirectPojoClientTest)  Time elapsed: 0.017 s
greetUnaryRequestVoid(org.apache.dubbo.sample.tri.direct.TriDirectPojoClientTest)  Time elapsed: 0.008 s
overload(org.apache.dubbo.sample.tri.direct.TriDirectPojoClientTest)  Time elapsed: 0.014 s
greetException(org.apache.dubbo.sample.tri.direct.TriDirectPojoClientTest)  Time elapsed: 0.036 s  <<< FAILURE!
java.lang.AssertionError
	at org.junit.Assert.fail(Assert.java:87)
	at org.junit.Assert.fail(Assert.java:96)
	at org.apache.dubbo.sample.tri.BaseTriPojoClientTest.greetException(BaseTriPojoClientTest.java:116)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray2(ReflectionUtils.java:206)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:161)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:84)
	at org.apache.dubbo.test.runner.ClassloaderSurefireStarter.runSuitesInProcess(ClassloaderSurefireStarter.java:78)
	at org.apache.dubbo.test.runner.TestRunnerMain.runInJunit4(TestRunnerMain.java:233)
	at org.apache.dubbo.test.runner.TestRunnerMain.main(TestRunnerMain.java:156)

greetLong(org.apache.dubbo.sample.tri.direct.TriDirectPojoClientTest)  Time elapsed: 1.466 s
```

File: dubbo-common/src/main/java/org/apache/dubbo/common/logger/log4j/Log4jLogger.java [hunk 3]
```diff
@@ -131,4 +131,8 @@ public boolean isErrorEnabled() {
         return logger.isEnabledFor(Level.ERROR);
     }
 
+    // test purpose only
```
Comment (AlbumenJ) [inline dubbo-common/src/main/java/org/apache/dubbo/common/logger/log4j/Log4jLogger.java:134]: Useless?

File: dubbo-common/src/main/java/org/apache/dubbo/common/logger/log4j2/Log4j2Logger.java [hunk 4]
```diff
@@ -125,4 +125,9 @@ public boolean isWarnEnabled() {
     public boolean isErrorEnabled() {
         return logger.isErrorEnabled();
     }
+
+    // test purpose only
```
Comment (AlbumenJ) [inline dubbo-common/src/main/java/org/apache/dubbo/common/logger/log4j2/Log4j2Logger.java:129]: Useless?

File: dubbo-common/src/main/java/org/apache/dubbo/common/utils/NetUtils.java [hunk 5]
```diff
@@ -679,14 +679,43 @@ public static void setInterface(MulticastSocket multicastSocket, boolean preferI
...
+     */
+    public static boolean matchIpExpression(String pattern, String address) throws UnknownHostException {
+        if (address == null) {
+            return false;
+        }
+
+        String host = address;
+        int port = 0;
+        // only works for ipv4 address with 'ip:port' format
+        if (address.endsWith(":")) {
```
Comment (AlbumenJ) [inline dubbo-common/src/main/java/org/apache/dubbo/common/utils/NetUtils.java:697]: Should be `contains`?

File: dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/integration/DynamicDirectory.java [hunk 6]
```diff
@@ -82,7 +82,7 @@
     /**
      * Initialization at construction time, assertion not null, and always assign non-null value
      */
-    protected final URL directoryUrl;
+    protected URL directoryUrl;
```
Comment (AlbumenJ) [inline dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/integration/DynamicDirectory.java:85]: Add `volatile`

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/configurator/AbstractConfigurator.java [hunk 7]
```diff
@@ -41,12 +45,15 @@
...
+import static org.apache.dubbo.rpc.cluster.Constants.RULE_VERSION_V30;
+import static org.apache.dubbo.rpc.cluster.configurator.parser.model.ConfiguratorConfig.MATCH_CONDITION;
 
 /**
  * AbstractConfigurator
  */
 public abstract class AbstractConfigurator implements Configurator {
-
+    public static final ErrorTypeAwareLogger errorLogger = LoggerFactory.getErrorTypeAwareLogger(AbstractConfigurator.class);
+    public static final Logger logger = LoggerFactory.getLogger(AbstractConfigurator.class);
```
Comment (AlbumenJ) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/configurator/AbstractConfigurator.java:56]: remove this logger

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/configurator/AbstractConfigurator.java [hunk 8]
```diff
@@ -41,12 +45,15 @@
...
 import static org.apache.dubbo.rpc.cluster.Constants.OVERRIDE_PROVIDERS_KEY;
+import static org.apache.dubbo.rpc.cluster.Constants.RULE_VERSION_V30;
+import static org.apache.dubbo.rpc.cluster.configurator.parser.model.ConfiguratorConfig.MATCH_CONDITION;
 
 /**
  * AbstractConfigurator
  */
 public abstract class AbstractConfigurator implements Configurator {
-
+    public static final ErrorTypeAwareLogger errorLogger = LoggerFactory.getErrorTypeAwareLogger(AbstractConfigurator.class);
```
Comment (AlbumenJ) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/configurator/AbstractConfigurator.java:55]: should be private

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

**Review by AlbumenJ — APPROVED** [review]
LGTM

**Pull request merged**